### PR TITLE
DBZ-4027 configurable readonly snapshot source

### PR DESCRIPTION
**Pull request by ismailsimsek**
https://issues.redhat.com/browse/DBZ-4027

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/pipeline/signal/ExternalSignalThread.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * <ul>
+ * <li>{@code id STRING} - the unique identifier of the signal sent, usually UUID, can be used for deduplication</li>
+ * <li>{@code type STRING} - the unique logical name of the code executing the signal</li>
+ * <li>{@code data STRING} - the data in JSON format that are passed to the signal code
+ * </ul> 
+ */
+public interface ExternalSignalThread {
+    void start();
+
+    void seek(long signalOffset);
```
Comment (ismailsimsek) [inline debezium-core/src/main/java/io/debezium/pipeline/signal/ExternalSignalThread.java:20]: is this correct interface to start implementations with? seek method is mostly kafka related, for examlpe file based  ExternalSignal not sure of seek method required.

File: debezium-core/src/main/java/io/debezium/pipeline/signal/FileSignalThread.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import io.debezium.config.Field;
+import io.debezium.connector.common.RelationalBaseSourceConnector;
+import io.debezium.pipeline.source.snapshot.incremental.AbstractReadOnlyIncrementalSnapshotChangeEventSource;
+import io.debezium.schema.DataCollectionId;
+
+/**
+ * The class responsible for processing of signals delivered to Debezium via a file.
+ * The signal message must have the following structure:
+ * <ul>
+ * <li>{@code data STRING} - the data in JSON format that are passed to the signal code
```
Comment (ismailsimsek) [inline debezium-core/src/main/java/io/debezium/pipeline/signal/FileSignalThread.java:31]: this is what i thought for file based signals. without `id` and `type`. should it be with`id`,`type`,`data`? comma separated?

```
{"data-collections": ["public.MyFirstTable", "public.MySecondTable"]}
{"data-collections": ["public.MyFirstTable", "public.MySecondTable2"]}
{"data-collections": ["public.MyFirstTable", "public.MySecondTable3"]}
{"data-collections": ["public.MyFirstTable", "public.MySecondTable4"]}
```

File: debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractReadOnlyIncrementalSnapshotChangeEventSource.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                                                                DataChangeEventListener dataChangeEventListener) {
+        super(config, jdbcConnection, dispatcher, databaseSchema, clock, progressListener, dataChangeEventListener);
+
+        // externalSignal = new KafkaSignalThread<>(MySqlConnector.class, config, this);
+        // dynamically load signaling source
+        // @TODO do we need Factory class here?
+        final String externalSignalClassName = config.getConfig().getString("signal.class", "io.debezium.pipeline.signal.KafkaSignalThread");
+        try {
+            // @TODO not sure this is correct way to get connectorType
+            final String connectorClassName = config.getConfig().getString("connector.class");
```
Comment (ismailsimsek) [inline debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractReadOnlyIncrementalSnapshotChangeEventSource.java:46]: i believe we need to get connector class here to instantiate signal thread here?

**Review by ismailsimsek — COMMENTED** [review]
Thank you @kgalieva im still trying to understand it please feel free to comment if its going to wrong direction. i left few questions maybe you can help with.

one general question: is it possible to use external source for non read-only snapshots? like using an external signaling table?

**Comment by kgalieva** [comment]
> one general question: is it possible to use external source for non read-only snapshots? like using an external signaling table?

no, it's not possible. The non read-only implementation counts on `window-open` and `window-close` watermarks to be in the same binlog stream as the rest of the binlog events. So it needs the signals table to be in the same database.

The `execute-snapshot `signal is replaceable with the  external signalling table, but you need to add the logic to non read-only implementation to periodically check if the new signals are available

File: debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java [hunk 4]
```diff
@@ -622,6 +631,10 @@ public String getLogicalName() {
 
     public abstract String getConnectorName();
 
+    public Class<RelationalBaseSourceConnector> getConnectorClass() throws ClassNotFoundException {
```
Comment (ismailsimsek) [inline debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:634]: not sure if this fits here?

**Comment by ismailsimsek** [comment]
@kgalieva its ready for review

**Comment by gunnarmorling** [comment]
Hey all, where are we with this one? IIUC, you wanted to give this a test ride, @kgalieva?

**Comment by gunnarmorling** [comment]
@ismailsimsek, could you rebase this one to the current upstream/main?

@kgalieva, are you still interested in giving this a test ride?

**Comment by Naros** [comment]
Hi all, just a gentle ping.  This PR has been dormant for just over 2 months and I'd like to see where we stand on this.

**Comment by ismailsimsek** [comment]
sorry for that, don't have time to work on this, closing

**Pull request closed without merging**